### PR TITLE
ODC-7519: Update the Pipeline Topology to adjust the new node design for the CustomRun

### PR DESCRIPTION
**Pull request by Lucifergene**
**Fixes**: 
https://issues.redhat.com/browse/ODC-7519

**Solution Description**: 

- [x] Design the new CustomTask node
- [x] Add the new icons to the console and use them as the Status Icons

**Screen shots / Gifs for design review**: 

Screenshot from 2024-02-16 19-16-07

Screenshot from 2024-02-16 19-16-28

Screenshot from 2024-02-16 19-16-49

https://github.com/openshift/console/assets/47265560/61e94a9e-3d8f-4fad-b450-0b6c74e557d5

**Unit test coverage report**: 
Not Changed.

**Test setup:**
1. Install Pipelines operator and Install the ApprovalTask Controller following this [Doc](https://github.com/openshift-pipelines/manual-approval-gate/blob/e1cf38bb54fa379c0efe271aec26ab7fa3fe4227/docs/DEVELOPEMENT.md).
2. Execute the following gist: https://gist.github.com/Lucifergene/f491504d7eda8a97e86a132029a9bc7b

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by MariaLeonova** [comment]
@Lucifergene that looks good! From the visual POV I'd suggest having less padding on the sides of the run bubble itself - right now there seems to be a bit of extra space. I'm adding the sizes from the figma file:
Screenshot 2024-02-19 at 9 53 10
So 8 px should be a good amount.

**Comment by Lucifergene** [comment]
> @Lucifergene that looks good! From the visual POV I'd suggest having less padding on the sides of the run bubble itself - right now there seems to be a bit of extra space. I'm adding the sizes from the figma file: Screenshot 2024-02-19 at 9 53 10 So 8 px should be a good amount.

Thanks @MariaLeonova! I have planned to adjust the paddings and spacings once the partially approved icons are implemented as part of the work for the next story. This will prevent any duplicate effort as part of the final touch-ups.

**Comment by sanketpathak** [comment]
@MariaLeonova @Lucifergene The UI here for the `wait` task(approval task) doesn't give any idea to the user to approve or disapprove the task. It looks ambiguous. 
Screenshot 2024-02-19 at 7 41 57 PM

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+  const { pipeline, pipelineRun, task } = element.getData();
+
+  const customTaskName = `${pipelineRun?.metadata?.name}-${task?.name}`;
+  const pipelineRunStatus = pipelineRun && pipelineRunFilterReducer(pipelineRun);
+  let customTaskStatus: string = '';
+
+  const watchedResources = {
+    customRun: {
+      kind: referenceForModel(CustomRunModelV1Beta1),
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx:177]: avoid using `referenceForModel` as it is deprecated.
```suggestion
       groupVersionKind: getGroupVersionKindForModel(CustomRunModelV1Beta1)
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+      name: customTaskName,
+      namespace: pipeline?.metadata?.namespace,
+      prop: 'task',
+    },
+    approvalTask: {
+      kind: referenceForModel(ApprovalTaskModel),
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx:183]:
```suggestion
      groupVersionKind: getGroupVersionKindForModel(ApprovalTaskModel),
```

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+import { Tooltip } from '@patternfly/react-core';
+import { observer, Node, NodeModel, useHover, createSvgIdUrl } from '@patternfly/react-topology';
+import * as cx from 'classnames';
+import * as _ from 'lodash';
+import { useTranslation } from 'react-i18next';
+import { Link } from 'react-router-dom-v5-compat';
+import { K8sResourceKind, WatchK8sResults } from '@console/dynamic-plugin-sdk/src/lib-core';
+import { resourcePathFromModel, truncateMiddle } from '@console/internal/components/utils';
+import { useK8sWatchResources } from '@console/internal/components/utils/k8s-watch-hook';
+import { referenceForModel } from '@console/internal/module/k8s';
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx:11]: avoid using `referenceForModel` as it is deprecated. Use `getGroupVersionKindForModel` as I suggested below.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+import { TaskNodeModelData } from './types';
+
+import './CustomTaskNode.scss';
+
+type ApprovalTaskNodeProps = {
+  element: Node<NodeModel, TaskNodeModelData>;
+  disableTooltip?: boolean;
+};
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx:34]: is it disable lint comment needed? I think you can remove it.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+  namespace: string;
+  disableVisualizationTooltip?: boolean;
+  width: number;
+  height: number;
+  customTask?: K8sResourceKind;
+}
+
+const FILTER_ID = 'SvgTaskDropShadowFilterId';
+
+const TaskComponent: React.FC<TaskProps> = ({
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx:56]: Better to name it `ApprovalTaskComponent`.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+      {visualName !== truncatedVisualName && disableVisualizationTooltip ? (
+        <Tooltip triggerRef={nameRef} content={visualName}>
+          {renderVisualName}
+        </Tooltip>
+      ) : (
+        renderVisualName
+      )}
+
+      {showStatusState && (
+        <>
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx:129]: nit: unnecessary fragment. You can remove it.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            <ApprovalStatusIcon status={status} />
+          </svg>
+        </>
+      )}
+    </g>
+  );
+
+  if (!disableVisualizationTooltip) {
+    taskPill = (
+      <>
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/ApprovalTaskNode.tsx:147]: nit: unnecessary fragment. You can remove it.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/CustomTaskNode.tsx [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+  const { height, width } = element.getBounds();
+
+  const { pipeline, pipelineRun, task } = element.getData();
+
+  const customTaskName = `${pipelineRun?.metadata?.name}-${task?.name}`;
+  const customTaskStatus: string = '';
+
+  const watchedResources = {
+    customRun: {
+      kind: referenceForModel(CustomRunModelV1Beta1),
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/CustomTaskNode.tsx:172]: use `getGroupVersionKindForModel` as mentioned above.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/CustomTaskNode.tsx [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+      {visualName !== truncatedVisualName && disableVisualizationTooltip ? (
+        <Tooltip triggerRef={nameRef} content={visualName}>
+          {renderVisualName}
+        </Tooltip>
+      ) : (
+        renderVisualName
+      )}
+
+      {showStatusState && (
+        <>
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/CustomTaskNode.tsx:125]: nit: remove fragment.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/CustomTaskNode.tsx [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            {<QuestionCircleIcon />}
+          </svg>
+        </>
+      )}
+    </g>
+  );
+
+  if (!disableVisualizationTooltip) {
+    taskPill = (
+      <>
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/CustomTaskNode.tsx:143]: nit: remove fragment.

File: frontend/packages/pipelines-plugin/src/utils/pipeline-augment.ts [hunk 11]
```diff
@@ -131,6 +138,21 @@ export const getRunStatusColor = (status: string): StatusMessage => {
   }
 };
 
+export const getApprovalStatusColor = (status: string): StatusMessage => {
+  switch (status) {
+    case ApprovalStatus.RequestSent:
+      return { message: i18next.t('pipelines-plugin~Request Sent'), pftoken: skippedColor };
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/utils/pipeline-augment.ts:144]:
```suggestion
      return { message: i18next.t('pipelines-plugin~Request sent'), pftoken: skippedColor };
```

File: frontend/packages/pipelines-plugin/src/utils/pipeline-augment.ts [hunk 12]
```diff
@@ -131,6 +138,21 @@ export const getRunStatusColor = (status: string): StatusMessage => {
...
+    case ApprovalStatus.Accepted:
+      return { message: i18next.t('pipelines-plugin~Approved'), pftoken: successColor };
+    case ApprovalStatus.Rejected:
+      return { message: i18next.t('pipelines-plugin~Rejected'), pftoken: failureColor };
+    case ApprovalStatus.TimedOut:
+      return { message: i18next.t('pipelines-plugin~Timed Out'), pftoken: cancelledColor };
```
Comment (vikram-raj) [inline frontend/packages/pipelines-plugin/src/utils/pipeline-augment.ts:150]:
```suggestion
      return { message: i18next.t('pipelines-plugin~Timed out'), pftoken: cancelledColor };
```

**Review by vikram-raj — COMMENTED** [review]
/lgtm
/approve

**Comment by Lucifergene** [comment]
> @MariaLeonova @Lucifergene The UI here for the `wait` task(approval task) doesn't give any idea to the user to approve or disapprove the task. It looks ambiguous. Screenshot 2024-02-19 at 7 41 57 PM

This has not been finalized yet. It will be added in the next stories.

**Comment by MariaLeonova** [comment]
@Lucifergene the screenshot link here doesn't work, can you please update? https://github.com/openshift/console/pull/13613#issuecomment-1985453164

**Comment by sanketpathak** [comment]
Verified with admin and non-admin users. Edge cases like affect on approval task after cancellation of pipeline run will be dealt later

https://github.com/openshift/console/assets/10252230/04d047b7-001e-42aa-926f-d1cb50545858

As tests will be added separately
/label qe-approved

**Review by vikram-raj — COMMENTED** [review]
/lgtm
